### PR TITLE
Adding support of AMDLIBM vector library

### PR DESCRIPTION
**Pull request by rohitaggarwal007**
Hi,

AMD has it's own implementation of vector calls. This patch include the changes to enable the use of AMD's math library using -fveclib=AMDLIBM.

### CODE REVIEW

**Comment by rohitaggarwal007** [comment]
@florianhumblot @alexey-bataev @RKSimon @phoebewang 
Please review the pull request.

File: llvm/include/llvm/Analysis/TargetLibraryInfo.h [hunk 1]
```diff
@@ -129,7 +129,8 @@ class TargetLibraryInfoImpl {
     MASSV,            // IBM MASS vector library.
     SVML,             // Intel short vector math library.
     SLEEFGNUABI, // SLEEF - SIMD Library for Evaluating Elementary Functions.
-    ArmPL        // Arm Performance Libraries.
+    ArmPL,       // Arm Performance Libraries.
+    AMDLIBM
```
Comment (phoebewang) [inline llvm/include/llvm/Analysis/TargetLibraryInfo.h:133]: Add comment?

File: clang/include/clang/Driver/Options.td [hunk 2]
```diff
@@ -3190,10 +3190,10 @@ def fno_experimental_isel : Flag<["-"], "fno-experimental-isel">, Group<f_clang_
 def fveclib : Joined<["-"], "fveclib=">, Group<f_Group>,
   Visibility<[ClangOption, CC1Option, FlangOption, FC1Option]>,
     HelpText<"Use the given vector functions library">,
-    Values<"Accelerate,libmvec,MASSV,SVML,SLEEF,Darwin_libsystem_m,ArmPL,none">,
+    Values<"Accelerate,libmvec,MASSV,SVML,SLEEF,Darwin_libsystem_m,ArmPL,AMDLIBM,none">,
```
Comment (phoebewang) [inline clang/include/clang/Driver/Options.td:3193]: Fix the format here

Comment (rohitaggarwal007) [reply clang/include/clang/Driver/Options.td:3193]: I did not understood the comment. you meant to say formatting?

File: llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls-finite.ll [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+; __FINITE_MATH_ONLY__ enabled, causing use of __<func>_finite
+; function versions, vectorization can map these to vector versions.
+
+target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-unknown-linux-gnu"
+
+declare float @__expf_finite(float) #0
+
+; CHECK-LABEL: @exp_f32
+; CHECK: <4 x float> @amd_vrs4_expf
```
Comment (RKSimon) [inline llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls-finite.ll:13]: Add RUNs for avx2 / avx512 capable targets to ensure amd_vrs8_expf / amd_vrs16_expf etc. are used when appropriate?

File: clang/include/clang/Driver/Options.td (hunk 2 above)
Comment (phoebewang) [reply clang/include/clang/Driver/Options.td:3193]: Sorry, it's some display problem in my side. I took it adding one more blank during the review.

File: llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls-finite.ll (hunk 3 above)
Comment (rohitaggarwal007) [reply llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls-finite.ll:13]: Added RUN for avx512 in amdlibm-calls.ll

File: llvm/lib/Analysis/TargetLibraryInfo.cpp [hunk 4]
```diff
@@ -1279,6 +1281,213 @@ void TargetLibraryInfoImpl::addVectorizableFunctionsFromVecLib(
     }
     break;
   }
+  case AMDLIBM: {
+#define FIXED(NL) ElementCount::getFixed(NL)
+    const VecDesc VecFuncs[] = {
```
[the remark this answers is not in the document]
Comment (rohitaggarwal007) [reply llvm/lib/Analysis/TargetLibraryInfo.cpp:1285]: done

File: llvm/include/llvm/Analysis/TargetLibraryInfo.h (hunk 1 above)
Comment (rohitaggarwal007) [reply llvm/include/llvm/Analysis/TargetLibraryInfo.h]: done

**Review by phoebewang — APPROVED** [review]
LGTM.

File: llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls-finite.ll (hunk 3 above)
Comment (RKSimon) [reply llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls-finite.ll:13]: I'm still not seeing cases where the vector width is varying depending on cpu abilities?

**Pull request merged**